mmgen/models/architectures/sngan_proj/generator_discriminator.py

```diff
@@ -0,0 +1,296 @@
+from copy import deepcopy
+
+import torch
+import torch.nn as nn
+from mmcv.cnn import ConvModule, build_activation_layer, xavier_init
+from mmcv.runner import load_checkpoint
+from mmcv.utils import build_from_cfg
+from torch.nn.utils import spectral_norm
+
+from mmgen.models.builder import MODULES
+from mmgen.utils import check_dist_init
+from mmgen.utils.logger import get_root_logger
+from ..common import get_module_device
+
+
+@MODULES.register_module()
+class SNGANGenerator(nn.Module):
+
+    # TODO: add doc here
+    _defualt_channels_cfg = {
+        32: [1, 1, 1],
+        64: [16, 8, 4, 2],
+        128: [16, 16, 8, 8, 4, 2]
+    }
+
+    _default_blocks_cfg = dict(type='SNGANGenResBlock')
+
+    def __init__(self,
+                 output_scale,
+                 noise_size,
+                 num_classes=0,
+                 out_channels=3,
+                 base_channels=128,
+                 input_scale=4,
+                 blocks_cfg=None,
+                 channel_cfg=None,
+                 auto_sync_bn=True,
+                 act_cfg=dict(type='ReLU')):
+        """this is a docstring
+        TODO:
+        """
+        super().__init__()
+
+        self.input_scale = input_scale
+        self.output_scale = output_scale
+        self.noise_size = noise_size
+
+        self.blocks_cfg = deepcopy(self._default_blocks_cfg)
+        if blocks_cfg is not None:
+            self.blocks_cfg.update(blocks_cfg)
+        if 'auto_sync_bn' not in self.blocks_cfg:
+            self.blocks_cfg['auto_sync_bn'] = auto_sync_bn
+        self.blocks_cfg['num_classes'] = num_classes
+        self.blocks_cfg['act_cfg'] = act_cfg
```

Comment:
For this part, you overwrite the keys if the keys have been set by the users. I recommend another way by adopting `setdefault`:

```python
self.blocks_cfg.setdefault(dict(num_classes=num_classes, act_cfg=act_cfg, auto_sync_bn=auto_sync_bn))
```